gitpod-io/gitpod
# [content-service] Refactor upload to GCS

## Description

Remove creation of chunks and CRC from the upload of backups. This is one the sources of CPU/RAM utilization in ws-daemon

## How to test

- Check tests
```
cd test && \
go test -timeout 300s \
  -run ^TestObjectUpload$ github.com/gitpod-io/gitpod/content-service/pkg/storage -v
```
or use docker
```
go test -timeout 300s \
  -run ^TestObjectUpload$ github.com/gitpod-io/gitpod/content-service/pkg/storage -v -args -with-docker
```

```
=== RUN   TestObjectUpload
=== RUN   TestObjectUpload/valid_1M_backup
 - - [06/May/2022:15:55:09 -0400] "POST /storage/v1/b?alt=json&prettyPrint=false&project=fake-project HTTP/1.1" 200 161
 - - [06/May/2022:15:55:09 -0400] "GET /storage/v1/b/gitpod-dev-user-fake-owner/o/workspaces%2Ffake-workspace%2Ffull.tar?alt=json&prettyPrint=false&projection=full HTTP/1.1" 404 59
 - - [06/May/2022:15:55:09 -0400] "POST /upload/storage/v1/b/gitpod-dev-user-fake-owner/o?alt=json&name=workspaces%2Ffake-workspace%2Ffull.tar&prettyPrint=false&projection=full&uploadType=multipart HTTP/1.1" 200 536
 - - [06/May/2022:15:55:09 -0400] "GET /gitpod-dev-user-fake-owner/workspaces/fake-workspace/full.tar HTTP/1.1" 200 1001984
=== RUN   TestObjectUpload/valid_100M_backup
 - - [06/May/2022:15:55:09 -0400] "POST /storage/v1/b?alt=json&prettyPrint=false&project=fake-project HTTP/1.1" 200 161
 - - [06/May/2022:15:55:10 -0400] "GET /storage/v1/b/gitpod-dev-user-fake-owner/o/workspaces%2Ffake-workspace%2Ffull.tar?alt=json&prettyPrint=false&projection=full HTTP/1.1" 404 59
 - - [06/May/2022:15:55:10 -0400] "POST /upload/storage/v1/b/gitpod-dev-user-fake-owner/o?alt=json&name=workspaces%2Ffake-workspace%2Ffull.tar&prettyPrint=false&projection=full&uploadType=resumable HTTP/1.1" 200 337
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 474
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 474
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 474
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 474
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 474
 - - [06/May/2022:15:55:10 -0400] "POST /upload/resumable/0a59d243db09ba73ba636e0dc4c23eb3 HTTP/1.1" 200 538
 - - [06/May/2022:15:55:10 -0400] "GET /gitpod-dev-user-fake-owner/workspaces/fake-workspace/full.tar HTTP/1.1" 200 100001792
--- PASS: TestObjectUpload (0.97s)
    --- PASS: TestObjectUpload/valid_1M_backup (0.02s)
    --- PASS: TestObjectUpload/valid_100M_backup (0.95s)
PASS
ok      github.com/gitpod-io/gitpod/content-service/pkg/storage 0.979s
```

## Release Notes
```release-note
[content-service] Refactor upload to GCS
```
